### PR TITLE
executor: fix two bugs when enable chunk

### PR DESCRIPTION
**Pull request by jackysp**
These two bugs were found when I tried to reproduce https://github.com/pingcap/tidb/issues/5935 .
When enable chunk,
1. forget to set field length in show executor.
2. write column info before call `NextChunk`.

~~I'll try to confirm these could fix the above issue. And it also need test cases in this pr, so DNM.~~
PTAL @XuHuaiyu @coocood

### CODE REVIEW

**Comment by jackysp** [comment]
This pr can fix #5935

**Comment by shenli** [comment]
@jackysp Please fix the CI.

File: executor/show.go [hunk 1]
```diff
@@ -103,6 +103,17 @@ func (e *ShowExec) NextChunk(ctx context.Context, chk *chunk.Chunk) error {
 		if err != nil {
 			return errors.Trace(err)
 		}
+		for i := 0; e.result != nil && i < e.result.NumRows(); i++ {
```
Comment (zz-jason) [inline executor/show.go]:
```go
iter := chunk.NewIterator4Chunk(e.result)
for colIdx := 0; colIdx < e.Schema().Len(); colIdx++ {
    retType := e.Schema().Columns[colIdx].RetType
    if !types.IsTypeVarchar(retType.Tp) {
        continue
    }
    for row := iter.Begin(); row != iter.End(); row = iter.Next() {
        if valLen := len(row.GetString(colIdx)); retType.Flen < valLen {
            retType.Flen = valLen
        }
    }
}
```

Comment (zz-jason) [reply executor/show.go]: this version is far more cache-friendly.

File: server/conn.go [hunk 2]
```diff
@@ -909,12 +909,7 @@ func (cc *clientConn) handleFieldList(sql string) (err error) {
 func (cc *clientConn) writeResultset(ctx context.Context, rs ResultSet, binary bool, more bool) error {
 	defer terror.Call(rs.Close)
 	if cc.server.cfg.EnableChunk && rs.SupportChunk() {
-		columns := rs.Columns()
-		err := cc.writeColumnInfo(columns)
-		if err != nil {
-			return errors.Trace(err)
-		}
-		err = cc.writeChunks(ctx, rs, binary, more)
+		err := cc.writeChunks(ctx, rs, binary, more)
```
Comment (zz-jason) [inline server/conn.go]: why make this change ?

File: server/conn.go [hunk 3]
```diff
@@ -991,12 +986,23 @@ func (cc *clientConn) writeColumnInfo(columns []*ColumnInfo) error {
...
+	gotColumnInfo := false
 	for {
 		// Here server.tidbResultSet implements NextChunk method.
 		err := rs.NextChunk(ctx, chk)
 		if err != nil {
 			return errors.Trace(err)
 		}
+		if !gotColumnInfo {
+			// We need to call NextChunk before we get columns.
+			// Otherwise, we will get incorrect columns info.
```
Comment (zimulala) [inline server/conn.go]: Could we move this logic to line 988?

Comment (zz-jason) [reply server/conn.go]: I think old logic is more simple. @jackysp

Comment (jackysp) [reply server/conn.go]: I think it is the correct way to fix it. Which old logic @zz-jason ?

Comment (zimulala) [reply server/conn.go]: Sorry, I misread the comment.

**Comment by zimulala** [comment]
LGTM

File: server/conn.go (hunk 3 above)
Comment (zz-jason) [reply server/conn.go]: misread +1

**Pull request merged**